wgpu-core/src/device/queue.rs

```diff
@@ -271,6 +271,21 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         device.pending_writes.consume(stage);
         device.pending_writes.dst_buffers.insert(buffer_id);
 
+        // Ensure the overwritten bytes are marked as initialized so they don't need to be nulled prior to mapping or binding.
+        {
+            let dst = buffer_guard.get_mut(buffer_id).unwrap();
+            for range in dst
+                .uninitialized_ranges_in_range::<Vec<std::ops::Range<wgt::BufferAddress>>>(
```

Comment:
in all the places where `uninitialized_ranges_in_range` is called, it would be good to fix two things:
  1. it should be an iterator, not a vector, to avoid the heap
  2. whatever it iterates on would just do `mark_initialized` internally

I.e. we could build an iterator that returns elements as it marks them initialized. Somewhat similar to `drain(..)` that returns elements and removes them.